### PR TITLE
Implement self.prefetch for zabbix_host

### PR DESCRIPTION
**Pull request by baurmatt**
#### Pull Request (PR) description
This commits adds the following improvements to zabbix_host:
* Implement self.prefetch (This allows the use of `puppet resource
zabbix_host` as well as purging unmanaged zabbix_host resources)
* Implement proper setters for all properties
* Rename the `group` parameter to `groups` to allow multiple groups to
be specified

This is another change for #570.

### CODE REVIEW

File: manifests/agent.pp [hunk 1]
```diff
@@ -223,6 +226,7 @@
   $monitored_by_proxy                     = $zabbix::params::monitored_by_proxy,
   $agent_use_ip                           = $zabbix::params::agent_use_ip,
   $zbx_group                              = $zabbix::params::agent_zbx_group,
+  $zbx_groups                             = $zabbix::params::agent_zbx_groups,
```
Comment (bastelfreak) [inline manifests/agent.pp]: Please add a datatype for the new option.

**Comment by bastelfreak** [comment]
@baurmatt can you take a look at the rubocop issues and add some tests for it? Thanks for the great work!

File: manifests/agent.pp (hunk 1 above)
Comment (baurmatt) [reply manifests/agent.pp]: Done :)

**Comment by baurmatt** [comment]
@bastelfreak Thanks! :) I've fixed the rubocop issues. Will take a look at unit tests for the type/provider.

**Comment by baurmatt** [comment]
@bastelfreak I've implemented all tests for the custom type. But the provider seems to be out of my league. From my understanding, that would require a complete mock of the Zabbix API.

**Comment by bastelfreak** [comment]
Thanks for all the work!

**Comment by guzmanbraso** [comment]
Hey, this commit broke our environments as the parameter group of manifests/resources/agent.pp was not renamed:
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: no parameter named 'groups' (file: /etc/puppetlabs/code/environments/x/modules/zabbix/manifests/resources/agent.pp, line: 38) on Zabbix_host[x] (file: /etc/puppetlabs/code/environments/x/modules/zabbix/manifests/resources/agent.pp, line: 38) on node x

**Comment by baurmatt** [comment]
@guzmanbraso First of all, sorry that it broke your environment! I will look into it and open up a new PR with the fix soon.

**Comment by baurmatt** [comment]
Hey @guzmanbraso, so I've just tried to understand what the problem could be, but sadly failed. manifests/resources/agent.pp still has the `group` parameter additionally to the new `groups` parameter:

https://github.com/voxpupuli/puppet-zabbix/blob/580b484f672d8003a6821ecd95d8e93df49c8092/manifests/resources/agent.pp#L21-L22

Could you please provide more information about what the problem is?

**Comment by alexjfisher** [comment]
@guzmanbraso Are you running `puppet generate types` to prevent environment isolation issues?

To me, it looks like your puppetserver doesn't realise the `type` contains a new `groups` parameter.

**Comment by guzmanbraso** [comment]
@baurmatt, no need to be sorry, my message was not a complain, just wanted to share that information wiyh you guys.  Because $groups was defined with spec I missed when fast searching for it and took as granted that $group had to be renamed. My bad

@alexjfisher will check that out and get back to you, I  did not tested if it was an environment isolation issue, thought it was something more basic but clearly it's not. I'll test that and get back to you.

Thank you everyone for your support!

**Comment by alexjfisher** [comment]
There were similar errors reported when puppetlabs/apt gained a new parameter in one of its ruby types. See https://github.com/voxpupuli/puppet-zabbix/issues/569#issuecomment-455482094

**Comment by hferrag** [comment]
> 
> 
> Hey, this commit broke our environments as the parameter group of manifests/resources/agent.pp was not renamed:
> Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: no parameter named 'groups' (file: /etc/puppetlabs/code/environments/x/modules/zabbix/manifests/resources/agent.pp, line: 38) on Zabbix_host[x] (file: /etc/puppetlabs/code/environments/x/modules/zabbix/manifests/resources/agent.pp, line: 38) on node x

I had exactly the same problem as @guzmanbraso

**Comment by hferrag** [comment]
I had exactly the same problem as @guzmanbraso 
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: no parameter named 'groups' (file: /etc/puppetlabs/code/z/modules/zabbix/manifests/resources/agent.pp, line: 39) on Zabbix_host[z] (file: /etc/puppetlabs/code/z/modules/zabbix/manifests/resources/agent.pp, line: 39) on node hfe-hana2.noc.lan

Fixed after running `puppet generate types  --environment my_env`.

Thanks @baurmatt and @alexjfisher  for your support.

**Pull request merged**